elastic/kibana
# [Metrics UI] Fix No Data alerts on Metric Threshold with groupBy

## Summary

Fixes https://github.com/elastic/kibana/issues/111772

This fixes the **Alert me when there's no data** option on a Metric Threshold alert that uses an **Alert per** setting. Any previously detected groups will be stored in the rule state, and if they fail to be detected on the next execution, they'll be backfilled with `null` values so that they can report no data.

**UPDATE:** Document count aggregated values will be backfilled with `0` instead, fixing the behavior of `doc_count <= 0` rules.

### Testing

The integration test should cover this scenario, but to ensure it works with the actual Kibana stack:

1. Start ingesting data from Metricbeat. Ideally with multiple nodes, but one node is sufficient to ensure this is working.
2. Create a metric threshold alert rule with a `groupBy` value, e.g. `host.name`, and turn on Alert Me When There's No Data. Add a server log action for the Alert state. Either set a threshold that will **not** fire, or turn off the Alert on Status Change throttling, to ensure that you'll see a No Data alert (as No Data and Alert use the same status).
3. Stop ingesting data. If you're ingesting data with multiple nodes, turn off only **some** of them, if possible.
4. Ensure that you get a No Data alert from only the nodes you stopped ingesting data from.

### Checklist
